core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -817,29 +817,49 @@ class Partition(val topicPartition: TopicPartition,
       case None => localReplica.logEndOffset.messageOffset
     }
 
-    // Only actually check the HW if this is a "latest" offset client request
-    if (isolationLevel.isDefined && timestamp == ListOffsetRequest.LATEST_TIMESTAMP) {
-      // Throw if the HW has not caught up with the start offset from this epoch
-      leaderEpochStartOffsetOpt
-        .filter(leo => leo > localReplica.highWatermark.messageOffset)
-        .map(leo => Errors.OFFSET_NOT_AVAILABLE.exception(s"Failed to fetch offsets for " +
-          s"partition $topicPartition with leader epoch ${currentLeaderEpoch.get} as this partition's " +
-          s"high watermark (${localReplica.highWatermark.messageOffset}) is lagging behind the " +
-          s"LEO at the start of this epoch ($leo)."))
-        .foreach(apiError => throw apiError)
-    }
-
-    if (timestamp == ListOffsetRequest.LATEST_TIMESTAMP) {
-      Some(new TimestampAndOffset(RecordBatch.NO_TIMESTAMP, lastFetchableOffset, Optional.of(leaderEpoch)))
-    } else {
-      def allowed(timestampOffset: TimestampAndOffset): Boolean =
-        timestamp == ListOffsetRequest.EARLIEST_TIMESTAMP || timestampOffset.offset < lastFetchableOffset
+    val invalidEpochStartOffset: Option[Long] = leaderEpochStartOffsetOpt
```

Comment:
I've been trying to come up with an alternative name. Basically we're in a situation where the local high watermark is not trusted and this is an upper bound on the true high watermark. Maybe `maxTrueHighWatermark`?